File: local_examples/cmds_hash/ioredis/cmds-hash.js
// EXAMPLE: cmds_hash

// HIDE_START
import assert from 'node:assert';
import { Redis } from 'ioredis';

const redis = new Redis();
// HIDE_END

// REMOVE_START
await redis.del('myhash');
// REMOVE_END

// STEP_START hmget
await redis.hset('myhash', { field1: 'Hello', field2: 'World' });

const hmgetResult = await redis.hmget('myhash', 'field1', 'field2', 'nofield');
console.log(hmgetResult); // >>> ['Hello', 'World', null]
// STEP_END

// REMOVE_START
assert.deepEqual(hmgetResult, ['Hello', 'World', null]);
await redis.del('myhash');
// REMOVE_END

// STEP_START hlen
const hlenSet1 = await redis.hset('myhash', 'field1', 'Hello');
console.log(hlenSet1); // >>> 1

const hlenSet2 = await redis.hset('myhash', 'field2', 'World');
console.log(hlenSet2); // >>> 1

const hlenResult = await redis.hlen('myhash');
console.log(hlenResult); // >>> 2
// STEP_END

// REMOVE_START
assert.equal(hlenSet1, 1);
assert.equal(hlenSet2, 1);
assert.equal(hlenResult, 2);
await redis.del('myhash');
// REMOVE_END

// STEP_START hset
const hsetRes1 = await redis.hset('myhash', 'field1', 'Hello');
console.log(hsetRes1); // >>> 1

console.log(await redis.hget('myhash', 'field1')); // >>> Hello

const hsetRes2 = await redis.hset('myhash', { field2: 'Hi', field3: 'World' });
console.log(hsetRes2); // >>> 2

console.log(await redis.hget('myhash', 'field2')); // >>> Hi
console.log(await redis.hget('myhash', 'field3')); // >>> World

const hsetRes3 = await redis.hgetall('myhash');
console.log(hsetRes3);
// >>> { field1: 'Hello', field2: 'Hi', field3: 'World' }
// STEP_END

// REMOVE_START
assert.equal(hsetRes1, 1);
assert.equal(hsetRes2, 2);
assert.deepEqual(hsetRes3, { field1: 'Hello', field2: 'Hi', field3: 'World' });
await redis.del('myhash');
// REMOVE_END

// STEP_START hget
const hgetRes1 = await redis.hset('myhash', 'field1', 'foo');
console.log(hgetRes1); // >>> 1

const hgetRes2 = await redis.hget('myhash', 'field1');
console.log(hgetRes2); // >>> foo

// A field that does not exist reads back as null.
const hgetRes3 = await redis.hget('myhash', 'field2');
console.log(hgetRes3); // >>> null
// STEP_END

// REMOVE_START
assert.equal(hgetRes1, 1);
assert.equal(hgetRes2, 'foo');
assert.equal(hgetRes3, null);
await redis.del('myhash');
// REMOVE_END

// STEP_START hgetall
await redis.hset('myhash', { field1: 'Hello', field2: 'World' });

const hgetallRes = await redis.hgetall('myhash');
console.log(hgetallRes); // >>> { field1: 'Hello', field2: 'World' }
// STEP_END

// REMOVE_START
assert.deepEqual(hgetallRes, { field1: 'Hello', field2: 'World' });
await redis.del('myhash');
// REMOVE_END

// STEP_START hdel
const hdelRes1 = await redis.hset('myhash', 'field1', 'foo');
console.log(hdelRes1); // >>> 1

const hdelRes2 = await redis.hdel('myhash', 'field1');
console.log(hdelRes2); // >>> 1

// Deleting a field that is not there removes nothing.
const hdelRes3 = await redis.hdel('myhash', 'field2');
console.log(hdelRes3); // >>> 0
// STEP_END

// REMOVE_START
assert.equal(hdelRes1, 1);
assert.equal(hdelRes2, 1);
assert.equal(hdelRes3, 0);
await redis.del('myhash');
// REMOVE_END

// STEP_START hvals
await redis.hset('myhash', { field1: 'Hello', field2: 'World' });

// HVALS follows the hash's field order, which Redis does not promise, so sort.
const hvalsRes = await redis.hvals('myhash');
console.log(hvalsRes.sort()); // >>> [ 'Hello', 'World' ]
// STEP_END

// REMOVE_START
assert.deepEqual(hvalsRes.sort(), ['Hello', 'World']);
await redis.del('myhash');
// REMOVE_END

// STEP_START hexpire
await redis.hset('myhash', { field1: 'Hello', field2: 'World' });

const hexpireRes1 = await redis.hexpire('myhash', 10, 'FIELDS', 2, 'field1', 'field2');
console.log(hexpireRes1); // >>> [ 1, 1 ]

const hexpireRes2 = await redis.httl('myhash', 'FIELDS', 2, 'field1', 'field2');
console.log(hexpireRes2); // >>> [ 10, 10 ]

// -2 means the field does not exist.
const hexpireRes3 = await redis.hexpire('myhash', 10, 'FIELDS', 1, 'nonexistent');
console.log(hexpireRes3); // >>> [ -2 ]
// STEP_END

// REMOVE_START
assert.deepEqual(hexpireRes1, [1, 1]);
assert.deepEqual(hexpireRes3, [-2]);
assert.ok(hexpireRes2.every((ttl) => ttl > 0 && ttl <= 10));
await redis.del('myhash');
// REMOVE_END

// HIDE_START
redis.disconnect();
// HIDE_END

File: local_examples/cmds_hash/hiredis/cmds_hash.c
// EXAMPLE: cmds_hash

// STEP_START includes
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <hiredis/hiredis.h>

// REMOVE_START
// Fail loudly on a NULL or error reply. hiredis returns an error REPLY (not a
// connection error) for things like a bad command, and the examples would
// otherwise print a wrong value and still exit 0 — a green harness run that
// proves nothing. Kept in a REMOVE block so the published example stays plain.
#define CHECK_REPLY(r) do { \
    if ((r) == NULL || (r)->type == REDIS_REPLY_ERROR) { \
        printf("REDIS ERROR: %s\n", (r) ? (r)->str : "no reply from server"); \
        return 1; \
    } \
} while (0)
// REMOVE_END
// STEP_END

int main(int argc, char **argv) {
    // STEP_START connect
    redisContext *c = redisConnect("127.0.0.1", 6379);

    if (c == NULL || c->err) {
        if (c) {
            printf("Connection error: %s\n", c->errstr);
            redisFree(c);
        } else {
            printf("Connection error: can't allocate redis context\n");
        }
        return 1;
    }
    // STEP_END

    // REMOVE_START
    redisReply *cleanup1 = redisCommand(c, "DEL myhash");
    CHECK_REPLY(cleanup1);
    freeReplyObject(cleanup1);
    // REMOVE_END

    // STEP_START hmget
    redisReply *reply;

    // Set up hash with fields
    reply = redisCommand(c, "HSET %s %s %s %s %s",
        "myhash", "field1", "Hello", "field2", "World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    freeReplyObject(reply);

    // Get multiple fields at once
    reply = redisCommand(c, "HMGET %s %s %s %s",
        "myhash", "field1", "field2", "nofield");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END

    printf("HMGET myhash field1 field2 nofield:\n");
    for (size_t i = 0; i < reply->elements; i++) {
        if (reply->element[i]->type == REDIS_REPLY_NIL) {
            printf("  [%zu]: null\n", i);
        } else {
            printf("  [%zu]: %s\n", i, reply->element[i]->str);
        }
    }
    // >>> [0]: Hello
    // >>> [1]: World
    // >>> [2]: null
    // STEP_END

    // REMOVE_START
    if (reply->elements != 3) {
        printf("ASSERTION FAILED: Expected 3 elements, got %zu\n", reply->elements);
    }
    if (strcmp(reply->element[0]->str, "Hello") != 0) {
        printf("ASSERTION FAILED: Expected 'Hello', got '%s'\n", reply->element[0]->str);
    }
    if (strcmp(reply->element[1]->str, "World") != 0) {
        printf("ASSERTION FAILED: Expected 'World', got '%s'\n", reply->element[1]->str);
    }
    if (reply->element[2]->type != REDIS_REPLY_NIL) {
        printf("ASSERTION FAILED: Expected nil for nofield\n");
    }
    // REMOVE_END

    freeReplyObject(reply);

    // REMOVE_START
    redisReply *cleanup2 = redisCommand(c, "DEL myhash");
    CHECK_REPLY(cleanup2);
    freeReplyObject(cleanup2);
    // REMOVE_END

    // STEP_START hlen
    // Add two new fields to the hash
    reply = redisCommand(c, "HSET %s %s %s", "myhash", "field1", "Hello");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("HSET myhash field1 Hello: %lld\n", reply->integer); // >>> 1
    // REMOVE_START
    if (reply->integer != 1) {
        printf("ASSERTION FAILED: Expected 1, got %lld\n", reply->integer);
    }
    // REMOVE_END
    freeReplyObject(reply);

    reply = redisCommand(c, "HSET %s %s %s", "myhash", "field2", "World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("HSET myhash field2 World: %lld\n", reply->integer); // >>> 1
    // REMOVE_START
    if (reply->integer != 1) {
        printf("ASSERTION FAILED: Expected 1, got %lld\n", reply->integer);
    }
    // REMOVE_END
    freeReplyObject(reply);

    // Count the fields in the hash
    reply = redisCommand(c, "HLEN %s", "myhash");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("HLEN myhash: %lld\n", reply->integer); // >>> 2
    // REMOVE_START
    if (reply->type != REDIS_REPLY_INTEGER) {
        printf("ASSERTION FAILED: Expected an integer reply for HLEN\n");
    }
    if (reply->integer != 2) {
        printf("ASSERTION FAILED: Expected 2, got %lld\n", reply->integer);
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    redisReply *del_reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(del_reply);
    freeReplyObject(del_reply);
    // REMOVE_END


    // STEP_START hset
    reply = redisCommand(c, "HSET myhash field1 Hello");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "HGET myhash field1");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> Hello
    freeReplyObject(reply);

    reply = redisCommand(c, "HSET myhash field2 Hi field3 World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 2
    freeReplyObject(reply);

    reply = redisCommand(c, "HGET myhash field2");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> Hi
    freeReplyObject(reply);

    reply = redisCommand(c, "HGET myhash field3");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> World
    freeReplyObject(reply);

    // HGETALL returns field and value alternating in a flat array.
    reply = redisCommand(c, "HGETALL myhash");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->elements; i += 2) {
        printf("%s=%s\n", reply->element[i]->str, reply->element[i + 1]->str);
    }
    // >>> field1=Hello
    // >>> field2=Hi
    // >>> field3=World
    // REMOVE_START
    if (reply->elements != 6) {
        printf("ASSERTION FAILED: Expected 6 entries, got %zu\n", reply->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START hget
    reply = redisCommand(c, "HSET myhash field1 foo");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "HGET myhash field1");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->str);
    // >>> foo
    freeReplyObject(reply);

    // A field that does not exist gives a nil reply, not an empty string.
    reply = redisCommand(c, "HGET myhash field2");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%s\n", reply->type == REDIS_REPLY_NIL ? "(nil)" : reply->str);
    // >>> (nil)
    // REMOVE_START
    if (reply->type != REDIS_REPLY_NIL) {
        printf("ASSERTION FAILED: Expected nil for a missing field\n");
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START hgetall
    reply = redisCommand(c, "HSET myhash field1 Hello field2 World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    freeReplyObject(reply);

    reply = redisCommand(c, "HGETALL myhash");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->elements; i += 2) {
        printf("%s=%s\n", reply->element[i]->str, reply->element[i + 1]->str);
    }
    // >>> field1=Hello
    // >>> field2=World
    // REMOVE_START
    if (reply->elements != 4) {
        printf("ASSERTION FAILED: Expected 4 entries, got %zu\n", reply->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START hdel
    reply = redisCommand(c, "HSET myhash field1 foo");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "HDEL myhash field1");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 1
    freeReplyObject(reply);

    // Deleting a field that is not there removes nothing.
    reply = redisCommand(c, "HDEL myhash field2");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->integer);
    // >>> 0
    // REMOVE_START
    if (reply->integer != 0) {
        printf("ASSERTION FAILED: Expected 0, got %lld\n", reply->integer);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START hvals
    reply = redisCommand(c, "HSET myhash field1 Hello field2 World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    freeReplyObject(reply);

    reply = redisCommand(c, "HVALS myhash");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->elements; i++) {
        printf("%s\n", reply->element[i]->str);
    }
    // >>> Hello
    // >>> World
    // REMOVE_START
    if (reply->elements != 2) {
        printf("ASSERTION FAILED: Expected 2 values, got %zu\n", reply->elements);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START hexpire
    reply = redisCommand(c, "HSET myhash field1 Hello field2 World");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    freeReplyObject(reply);

    // HEXPIRE needs the field count before the field names.
    reply = redisCommand(c, "HEXPIRE myhash 10 FIELDS 2 field1 field2");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->elements; i++) {
        printf("%lld\n", reply->element[i]->integer);
    }
    // >>> 1
    // >>> 1
    freeReplyObject(reply);

    reply = redisCommand(c, "HTTL myhash FIELDS 2 field1 field2");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    for (size_t i = 0; i < reply->elements; i++) {
        printf("%lld\n", reply->element[i]->integer);
    }
    // >>> 10
    // >>> 10
    // REMOVE_START
    for (size_t i = 0; i < reply->elements; i++) {
        if (reply->element[i]->integer <= 0 || reply->element[i]->integer > 10) {
            printf("ASSERTION FAILED: Unexpected TTL %lld\n", reply->element[i]->integer);
            return 1;
        }
    }
    // REMOVE_END
    freeReplyObject(reply);

    // -2 means the field does not exist.
    reply = redisCommand(c, "HEXPIRE myhash 10 FIELDS 1 nonexistent");
    // REMOVE_START
    CHECK_REPLY(reply);
    // REMOVE_END
    printf("%lld\n", reply->element[0]->integer);
    // >>> -2
    // REMOVE_START
    if (reply->element[0]->integer != -2) {
        printf("ASSERTION FAILED: Expected -2, got %lld\n", reply->element[0]->integer);
        return 1;
    }
    // REMOVE_END
    freeReplyObject(reply);
    // STEP_END

    // REMOVE_START
    reply = redisCommand(c, "DEL myhash");
    CHECK_REPLY(reply);
    freeReplyObject(reply);
    // REMOVE_END

    // STEP_START disconnect
    redisFree(c);
    // STEP_END

    return 0;
}

